// @parity/abi/lib/util/types.js
"use strict";
// Copyright 2015-2019 Parity Technologies (UK) Ltd.
// This file is part of Parity.
//
// SPDX-License-Identifier: MIT
Object.defineProperty(exports, "__esModule", { value: true });
var HEXDIGITS = [
    '0',
    '1',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    'a',
    'b',
    'c',
    'd',
    'e',
    'f'
];
function isArray(input) {
    return Array.isArray(input);
}
exports.isArray = isArray;
function isError(input) {
    return input instanceof Error;
}
exports.isError = isError;
function isFunction(input) {
    return typeof input === 'function';
}
exports.isFunction = isFunction;
function isHex(input) {
    if (!isString(input)) {
        return false;
    }
    if (input.substr(0, 2) === '0x') {
        return isHex(input.slice(2));
    }
    var lowerCaseInput = input.toLowerCase();
    var hex = true;
    for (var index = 0; hex && index < input.length; index++) {
        hex = HEXDIGITS.includes(lowerCaseInput[index]);
    }
    return hex;
}
exports.isHex = isHex;
function isObject(input) {
    return Object.prototype.toString.call(input) === '[object Object]';
}
exports.isObject = isObject;
function isString(input) {
    return typeof input === 'string';
}
exports.isString = isString;
function isInstanceOf(input, clazz) {
    return input instanceof clazz;
}
exports.isInstanceOf = isInstanceOf;
