// @parity/api/lib/transport/jsonRpcEncoder.js
"use strict";
// Copyright 2015-2018 Parity Technologies (UK) Ltd.
// This file is part of Parity.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
var EventEmitter = require('eventemitter3');
var JsonRpcEncoder = /** @class */ (function (_super) {
    __extends(JsonRpcEncoder, _super);
    function JsonRpcEncoder() {
        var _this = _super.call(this) || this;
        _this._id = 1;
        return _this;
    }
    JsonRpcEncoder.prototype.encodeObject = function (method, params) {
        return {
            id: this._id++,
            jsonrpc: '2.0',
            method: method,
            params: params
        };
    };
    JsonRpcEncoder.prototype.encode = function (method, params) {
        return JSON.stringify(this.encodeObject(method, params));
    };
    Object.defineProperty(JsonRpcEncoder.prototype, "id", {
        get: function () {
            return this._id;
        },
        enumerable: true,
        configurable: true
    });
    return JsonRpcEncoder;
}(EventEmitter));
module.exports = JsonRpcEncoder;
