// @parity/api/lib/rpc/shell/shell.js
"use strict";
// Copyright 2015-2018 Parity Technologies (UK) Ltd.
// This file is part of Parity.
// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
var Shell = /** @class */ (function () {
    function Shell(provider) {
        this._provider = provider;
    }
    Shell.prototype.getApps = function (retrieveAll) {
        if (retrieveAll === void 0) { retrieveAll = false; }
        return this._provider
            .send('shell_getApps', retrieveAll);
    };
    Shell.prototype.getFilteredMethods = function () {
        return this._provider
            .send('shell_getFilteredMethods');
    };
    Shell.prototype.getMethodGroups = function () {
        return this._provider
            .send('shell_getMethodGroups');
    };
    Shell.prototype.getMethodPermissions = function () {
        return this._provider
            .send('shell_getMethodPermissions');
    };
    Shell.prototype.loadApp = function (appId, params) {
        if (params === void 0) { params = null; }
        return this._provider
            .send('shell_loadApp', appId, params);
    };
    Shell.prototype.setMethodPermissions = function (permissions) {
        return this._provider
            .send('shell_setMethodPermissions', permissions);
    };
    Shell.prototype.setAppVisibility = function (appId, state) {
        return this._provider
            .send('shell_setAppVisibility', appId, state);
    };
    Shell.prototype.setAppPinned = function (appId, state) {
        return this._provider
            .send('shell_setAppPinned', appId, state);
    };
    return Shell;
}());
module.exports = Shell;
